packages/world/ts/config/v2/system.ts

```diff
@@ -0,0 +1,82 @@
+import { SYSTEM_DEFAULTS } from "./defaults";
+import { SystemInput } from "./input";
+import { hasOwnKey, mergeIfUndefined } from "@latticexyz/store/config/v2";
+import { ErrorMessage, narrow, requiredKeyOf } from "@arktype/util";
+import { Hex } from "viem";
+import { resourceToHex } from "@latticexyz/common";
+
+export type ValidateSystemOptions = { readonly inNamespace?: true };
+
+export type requiredSystemKey<inNamespace extends true | undefined> = Exclude<
+  requiredKeyOf<SystemInput>,
+  inNamespace extends true ? "label" | "namespace" : never
+>;
+
+export type validateSystem<input, options extends ValidateSystemOptions = {}> = {
+  [key in keyof input | requiredSystemKey<options["inNamespace"]>]: key extends keyof SystemInput
+    ? key extends "label" | "namespace"
+      ? options["inNamespace"] extends true
+        ? ErrorMessage<"Overrides of `label` and `namespace` are not allowed for systems in this context.">
+        : key extends keyof input
+          ? narrow<input[key]>
+          : never // TODO: should this be SystemInput[key] ?
```

Comment:
i think `never` is good, this clause should never be reached in practice (since `input` will only have keys that appear in `input`)